server/routerlicious/packages/lambdas/src/utils/circuitBreaker.ts

```diff
@@ -0,0 +1,207 @@
+	private readonly context: IContext;
+	private readonly dependencyName: string;
+	private readonly executeFunction: (...args: any[]) => Promise<any>;
+	private readonly healthCheckFunction: (...args: any[]) => Promise<any>;
+	private readonly healthCheckParams: any[] = [];
+	private readonly fallbackToRestartTimeoutMs: number = 180000; // 3 minutes by default
+
+	// following properties are used for telemetry and reset when the circuit breaker is closed
+	private circuitBreakerMetric: Lumber<LumberEventName.CircuitBreaker> | undefined;
+	private circuitBreakerOpenCount: number = 0;
+	private error: any;
+	private fallbackToRestartTimeout: NodeJS.Timeout | undefined; // timeout to restart the service if the circuit breaker is not closed for more than fallbackToRestartTimeoutMs
+
+	constructor(
+		circuitBreakerOptions: circuitBreakerOptions,
+		context: IContext,
+		dependencyName: string,
+		executeFunction: (...args: any[]) => Promise<any>,
+		healthCheckFunction: (...args: any[]) => Promise<any>,
+		healthCheckParams?: any[],
+	) {
+		this.context = context;
+		this.dependencyName = dependencyName;
+		this.executeFunction = executeFunction;
+		this.healthCheckFunction = healthCheckFunction;
+		this.healthCheckParams = healthCheckParams ?? [];
+		this.fallbackToRestartTimeoutMs =
+			circuitBreakerOptions.fallbackToRestartTimeoutMs ?? this.fallbackToRestartTimeoutMs;
+
+		const lambdaErrorFilter = circuitBreakerOptions.errorFilter;
+
+		this.circuitBreaker = new CircuitBreaker(wrapperCircuitBreakerAction, {
+			errorThresholdPercentage: circuitBreakerOptions.errorThresholdPercentage,
+			resetTimeout: circuitBreakerOptions.resetTimeout,
+			timeout: circuitBreakerOptions.timeout,
+			rollingCountTimeout: circuitBreakerOptions.rollingCountTimeout,
+			rollingCountBuckets: circuitBreakerOptions.rollingCountBuckets,
+			errorFilter: (error): boolean => {
+				if (error.healthCheckFailed) {
+					// open the circuit breaker if health check fails with any error, else use lambdaErrorFilter
+					return false;
+				}
+				return lambdaErrorFilter ? lambdaErrorFilter(error) : false;
+			},
+		});
+
+		this.circuitBreaker.on("open", () => this.openCallback());
+		this.circuitBreaker.on("close", () => this.closeCallback());
+		this.circuitBreaker.on("halfOpen", () => this.halfOpenCallback());
+		this.circuitBreaker.on("failure", (err) => this.failureCallback(err)); // Emitted when the circuit breaker action fails
+		this.circuitBreaker.on("reject", (err) => this.rejectCallback(err)); // Emitted when the circuit breaker is in open state (failing fast) and action is fired
+	}
+
+	public async execute(params: any[], doHealthCheck?: boolean): Promise<void> {
+		const functionToFire =
+			doHealthCheck && this.healthCheckFunction
+				? this.healthCheckFunction
+				: this.executeFunction;
+		return this.circuitBreaker.fire(functionToFire, params); // calls wrapperCircuitBreakerAction with these params
+	}
+
+	public getCurrentState(): string {
```

Comment:
We can probably simplify this to `this.circuitBreaker.toJSON()?.state ?? "unknown"`

Reply:
circuitBreaker.toJSON().state returns an object of below type:
```
{
  enabled: true,
  name: 'functionName'
  closed: true,
  open: false,
  halfOpen: false,
  warmUp: false,
  shutdown: false
}
```

Hence directly using the circuitBreaker.opened / halfOpen instance members.